modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/Output.scala

```diff
@@ -12,30 +12,89 @@
 
 object Output {
 
-  case class Bad(badRow: BadRow) extends Output
-  case class Good(event: EnrichedEvent) extends Output
-  case class Pii(event: EnrichedEvent) extends Output
+  /** An item destined for the Bad Sink */
+  case class Bad[B](badRow: B) extends Output[B, Nothing, Nothing]
+
+  /** An item destined for the Good Sink */
+  case class Good[G](event: G) extends Output[Nothing, G, Nothing]
+
+  /** An item destined for the Pii Sink */
+  case class Pii[P](event: P) extends Output[Nothing, Nothing, P]
 
-  /** A variant of fs2's observeEither, sinking three output types into three Pipe's */
-  def sink[F[_]: Concurrent](
-    bad: Pipe[F, BadRow, Unit],
-    good: Pipe[F, EnrichedEvent, Unit],
-    pii: Pipe[F, EnrichedEvent, Unit]
-  ): Pipe[F, Output, Unit] =
+  type Parsed = Output[BadRow, EnrichedEvent, EnrichedEvent]
+  type Serialized = Output[Array[Byte], Array[Byte], Array[Byte]]
+
+  /** Sends a bad/good/pii item into the appropriate sink */
+  def sink[F[_]: Concurrent, B, G, P](
+    bad: Pipe[F, B, Unit],
+    good: Pipe[F, G, Unit],
+    pii: Pipe[F, P, Unit]
+  ): Pipe[F, Output[B, G, P], Unit] =
     _.observe(_.collect { case Bad(br) => br }.through(bad))
       .observe(_.collect { case Good(event) => event }.through(good))
       .observe(_.collect { case Pii(event) => event }.through(pii))
       .drain
 
+  /**
+   * Serializes events and bad rows into byte arrays. Good/Pii events exceeding
+   *  a maximum size after serialization are converted to bad events.
+   */
+  val serialize: Parsed => Serialized =
+    serializeGood.andThen(serializePii).andThen(serializeBad)
+
+  private def serializeGood[P]: Output[BadRow, EnrichedEvent, P] => Output[BadRow, Array[Byte], P] = {
+    case Good(g) =>
+      serializeEnriched(g).fold(Bad(_), Good(_))
+    case b: Bad[BadRow] => b
+    case p: Pii[P] => p
+  }
+
+  private def serializePii[G]: Output[BadRow, G, EnrichedEvent] => Output[BadRow, G, Array[Byte]] = {
+    case Pii(p) =>
+      serializeEnriched(p).fold(Bad(_), Pii(_))
+    case b: Bad[BadRow] => b
+    case g: Good[G] => g
+  }
+
+  private def serializeBad[G, P]: Output[BadRow, G, P] => Output[Array[Byte], G, P] = {
+    case Bad(br) => Bad(br.compact.getBytes(UTF_8))
+    case g: Good[G] => g
+    case p: Pii[P] => p
+  }
+
+  private def serializeEnriched(enriched: EnrichedEvent): Either[BadRow, Array[Byte]] = {
+    val asStr = ConversionUtils.tabSeparatedEnrichedEvent(enriched)
+    val asBytes = asStr.getBytes(UTF_8)
+    val size = asBytes.length
+    if (size > MaxRecordSize) {
+      val msg = "event passed enrichment but exceeded the maximum allowed size as a result"
+      val br = BadRow
+        .SizeViolation(
+          Enrich.processor,
+          Failure.SizeViolation(Instant.now(), MaxRecordSize, size, msg),
+          BadRowPayload.RawPayload(asStr.take(MaxMessageSize))
+        )
+      Left(br)
+    } else Right(asBytes)
+  }
+
+  private val MaxRecordSize = 6900000
```

Comment:
Also probably worth mentioning in a comment why it's `6900000` while official is something around 10Mb

Reply:
I will mention it is because of base64 encoding.

By the way, I would have guessed the max cut off would be closer to 7500000.  Because base64 encoding has a 4/3 expansion factor.